### PR TITLE
HDDS-5757. balancer should stop when the cluster can not be balanced any more

### PR DESCRIPTION
**Pull request by JacksonYao287**
## What changes were proposed in this pull request?

when i test container balancer in k8s cluster,  i use the command line :

./ozone admin containerbalancer start -i -1 -t 0.000001 -d 1 -s 500, and set the contaienr size to 1G.

i found that balancer thread can not stop when the cluster is close to balance.

in my cluster , i have three datanodes d1,d2,d3(disk usage are 67G, 67G, 68G), and the fourth d4 datanode's disk usage is 1G. when i start balancer, it begin to balance and work well, many containers have been moved from d1,d2,d3 to d4. but when the cluster is close to balance(the disk usages of the four datanodes are 50G,52G,50G,51G), the balancer is still running , it will move container among those datanodes again and again.

let us see a general example. if we have two datanode d1 and  d2,  the disk usage are 3G and 7G respectively, then the average usage is 5G. if  we set the threshold to 0.00001(close to 0), the lowerLimit and upperLimit is close to 5G. if we set the container size to 4G,  in the first iteraion, we can recognize the over-utilized datanode d2 and the under-utilized datanode d1, then we schedule a move option from d2 to d1 with a 4G container. after move option is finished , d1 is 7G and d2 is 3G. so the balancer thread will go on for ever.

 so we should let the balancer thread exit when the cluster can not be more balanced

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5757

### CODE REVIEW

**Comment by JacksonYao287** [comment]
@lokeshj1703 @siddhantsangwan PTAL, thanks!

**Review by lokeshj1703 — COMMENTED** [review]
@JacksonYao287 Thanks for working on this PR! 
I feel we can add extra check in `canSizeEnterTarget` function in ContainerBalancer. That might be sufficient for what we need.

**Comment by JacksonYao287** [comment]
> I feel we can add extra check in `canSizeEnterTarget` function in ContainerBalancer. That might be sufficient for what we need.
@lokeshj1703 make sense , thanks for the review! i will do this in new commit

**Review by siddhantsangwan — APPROVED** [review]
@JacksonYao287 changes look good! We can add more documentation around `isMoveGenerated` to `doIteration` and `canSizeEnterTarget` to make these changes clearer.

**Comment by JacksonYao287** [comment]
thanks @siddhantsangwan for the review , i have add more documentation in the new commit

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java [hunk 1]
```diff
@@ -652,8 +670,24 @@ private double ratioToBytes(Long nodeCapacity, double utilizationRatio) {
...
+
+      //size can be moved into target datanode only when the following
+      //two condition are met.
+      //1 sizeEnteringAfterMove does not succeed the configured
+      // MaxSizeEnteringTarget
+      //2 current usage of target datanode plus sizeEnteringAfterMove
+      // is smaller than or equal to upperLimit
+      double usedAfterMove =
+          capacity - scmNodeStat.getRemaining().get() + sizeEnteringAfterMove;
+
```
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:688]: Can we move this calculation in a separate Utils class or sth like that? @siddhantsangwan had added a function in DatanodeUsageInfo I think?
I feel we can reuse the same function. We will only need a parameter for used size to add.

**Review by lokeshj1703 — COMMENTED** [review]
@JacksonYao287 Thanks for updating the PR! I have few comments inline.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java (hunk 1 above)
Comment (siddhantsangwan) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:688]: `DatanodeUsageInfo.calculateUtilization`

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java [hunk 2]
```diff
@@ -187,7 +190,13 @@ private void balance() {
         stop();
         return;
       }
-      doIteration();
+
+      //if no new move option is generated, it means the cluster can
+      //not be balanced any more , so just stop
+      if (doIteration() == IterationResult.CAN_NOT_BALANCE_ANY_MORE) {
```
[the remark this answers is not in the document]
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:196]: sure, will do this in a new commit

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java (hunk 1 above)
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java]: yea , i have updated this patch , PTAL!

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/DatanodeUsageInfo.java [hunk 3]
```diff
@@ -60,25 +61,27 @@ private static int compareByUtilization(DatanodeUsageInfo first,
...
+   * by its capacity. Here, we prefer calculating
    * used space as (capacity - remaining), instead of using
    * {@link SCMNodeStat#getScmUsed()} (see HDDS-5728).
    *
+   * @param plusSize the increased size
    * @return (capacity - remaining) / capacity of this datanode
    */
-  public double calculateUtilization() {
-    double capacity = scmNodeStat.getCapacity().get();
+  public double calculateUtilization(long plusSize) {
```
Comment (lokeshj1703) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/DatanodeUsageInfo.java:78]: NIT: Let's have another function `calculateUtilization()` which calls this function with 0 as parameter. Then we can keep using the older function in other places.

**Review by lokeshj1703 — COMMENTED** [review]
@JacksonYao287 Thanks for updating the PR!  The changes look good to me. Just a minor comment.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/DatanodeUsageInfo.java (hunk 3 above)
Comment (JacksonYao287) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/DatanodeUsageInfo.java:78]: sure, i will do this

**Review by lokeshj1703 — APPROVED** [review]
@JacksonYao287 Thanks for updating the PR! The changes look good to me. +1.

**Comment by JacksonYao287** [comment]
@siddhantsangwan @lokeshj1703 thank you for the review!

**Pull request merged**